### ISSUE TITLE
Derivatives of `centroid`-interpolated variables

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#4854]
This very useful 2007 article about `centroid` shading says that - in OpenGL in 2007 - derivatives on centroid-interpolated variables have undefined results: https://www.opengl.org/pipeline/article/vol003_6/
> When you need accurate derivatives (explicit or implied, such as with mipmapped or anisotropic texture fetches).  The shading language specification considers derivatives derived from centroid varings to be so fraught with inaccuracy that it was resolved they are simply undefined.  In such a case, strongly consider  at least adding:      centroid varying float myMixer; // beware of derivative!

Vulkan is a bit less harsh than that, at least in this non-normative note (I couldn't find normative text):
> As Centroid restricts the possible interpolation position to the covered area of the primitive, the position can be forced to vary between neighboring fragments when it otherwise would not. Derivatives calculated based on these differing locations can produce inconsistent results compared to undecorated inputs. It is recommended that input variables used in derivative calculations are not decorated with Centroid.

FWIW, D3D is very strict about centroid interpolation, so it probably doesn't have issues with derivatives: https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#3.5.5%20Centroid%20Sampling%20of%20Attributes

But anyway, should WGSL say this is a dynamic error? Should it be a compile-time warning or error?

**Comment by alan-baker** [comment]
Am I interpreting correctly that it's not that the derivatives are wrong, but that the values of the inputs might not be what authors anticipate? If that's the case I wouldn't want to change derivatives, but it might make sense to add a note to centroid sampling that calls out the situation in some way.

**Comment by kainino0x** [comment]
That's my understanding of Vulkan, yes. I would be surprised if OpenGL is ever actually worse than that in practice, even if the spec allows that (or used to allow it - I didn't dig up the current OpenGL spec text).

**Comment by jimblandy** [comment]
This sounds like a note in the spec is sufficient?

**Comment by kdashg** [comment]
Yeah it's probably useful to as an example call out that "because centroids are not necessarily one pixel width apart, derivative functions might not have the values you expect".

**Comment by dneto0** [comment]
MSL and Vulkan wording is simliar to what exists in WGSL:

MSL: (5.4 sampling and interpolation attributes)

> If a centroid attribute variant is specified (centroid_perspective and
centroid_no_perspective), the interpolation point sampled needs to be within both the
primitive and the centroid of the pixel.

Vulkan: 9.12 interpolation decorations (the normative above the paragraph Kai quoted):

> If the Centroid decoration is used, the interpolation position used for the variable must also fall within the bounds of the primitive being rasterized.

WGSL:

> centroid:  Interpolation is performed at a point that lies within all the samples covered by the fragment within the current primitive. This value is the same for all samples in the primitive.

D3D is more prescriptive about exactly what sample location is chosen, while the others allow a lot more choices, potentially.

But all three have the "you're not likely to get something intuitive" vibe.   I'd go with a note at best.
